### ISSUE TITLE
Results are different on MacOS for fixed data

### ISSUE THREAD

**Issue by madrugado** [tensorflow/tensorflow#20308]
Hello.

We've encountered strange behaviour of TF (tested on versions 1.7.0 and 1.8.0) on MacOS (tested on 10.11.6) with Python 2.7 (tested on 2.7.10). While running this code:

```python
import tensorflow as tf

class Model(object):
    def __init__(self):
        one_hot = tf.constant([[1, 0, 0, 1], [0, 1, 1, 2], [1, 1, 1, 3]])
        logits = tf.constant([[0.3, 0.2, 0], [0.1, 0.0004, 0], [0, 0.02, 0.2]])

        self.one_hot = one_hot
        self.logits = logits
        self.rec_at_k = self._recall_at_k()
        self.prec_at_k = self._precision_at_k()

        config = tf.ConfigProto()
        config.gpu_options.allow_growth = True
        self.sess = tf.Session(config=config)
        self.sess.run(tf.global_variables_initializer())
        self.sess.run(tf.local_variables_initializer())


    def _recall_at_k(self, top=3):
        recalls = []
        for k in range(1,top+1):
            recalls.append(tf.metrics.recall_at_k(tf.cast(self.one_hot, tf.int64), tf.cast(self.logits, tf.float32), k=k))
        return recalls

    def _precision_at_k(self, top=3):
        precs = []
        for k in range(1,top+1):
            precs.append(tf.metrics.recall_at_k(tf.cast(self.one_hot, tf.int64), self.logits, k=k))
        return precs

    def fit(self):
        recalls, precs = self.sess.run([self.rec_at_k, self.prec_at_k])
        for rec,prec in zip(recalls,precs):
            print rec[0],rec, prec
        l,o = self.sess.run([self.logits, self.one_hot])
        print l
        print o, o[:,:-1]


model = Model()
model.fit()
```

we receive results which are different from run to run, like:

```
0.2857142857142857 ('0.2857142857142857', '0.2857142857142857') ('1.0', '0.2857142857142857')
0.7142857142857143 ('0.7142857142857143', '0.7142857142857143') ('1.0', '0.7142857142857143')
0.0 ('0.0', '0.8571428571428571') ('1.0', '0.8571428571428571')
[[0.3    0.2    0.    ]
 [0.1    0.0004 0.    ]
 [0.     0.02   0.2   ]]
[[1 0 0 1]
 [0 1 1 2]
 [1 1 1 3]] [[1 0 0]
 [0 1 1]
 [1 1 1]]
```

and

```
0.4 ('0.4', '0.2857142857142857') ('1.0', '0.2857142857142857')
1.0 ('1.0', '0.7142857142857143') ('0.7142857142857143', '0.7142857142857143')
0.0 ('0.0', '0.8571428571428571') ('1.0', '0.8571428571428571')
[[0.3    0.2    0.    ]
 [0.1    0.0004 0.    ]
 [0.     0.02   0.2   ]]
[[1 0 0 1]
 [0 1 1 2]
 [1 1 1 3]] [[1 0 0]
 [0 1 1]
 [1 1 1]]
```

On linux (Ubuntu 16.04 4.13.0-45-generic) and Python (2.7.14) results are stable.

**Comment by madrugado** [comment]
additionally tested on MacOS 10.13.3 with Python 2.7.14 Anaconda custom (64-bit)

and linux:
Ubuntu linux 4.4.0-64-generic with Python 2.7.12

**Comment by cy89** [comment]
This reminds me of #15933 (https://github.com/tensorflow/tensorflow/issues/15933). Any chance you're hitting the same thing, or a relative of it? I think the primary recommendation from that bug was to try at least 10.12.6 (which your latest comment suggests you've done).

**Comment by madrugado** [comment]
It seems to be kinda related, as far as I understand that issue, the problem is that two versions of tensorflow outputs different matrices in same conditions. I haven't tried to draw the output, but I think they are pretty different in a visual way too. 
In my case the output is changing from run to run, like some information is random and/or stored between the runs.

I've written above that I've tried different recent version of MacOS, Tensorflow and Python as well.

**Comment by cy89** [comment]
@madrugado within one version of TF, do you get self-consistent results? 
And, do either versions' results match Linux? Or do they both differ from Linux?

**Comment by madrugado** [comment]
Yes, I get self consistent results, I've tried to re-run the script multiple times on different macOS, with restart of mac.

The results on linux are also matched (and they are as expected in the first place).

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by cy89** [comment]
@quaeler @Gemesys @lewisl can we interest you in this bug?

**Comment by quaeler** [comment]
(sorry - i've not enough free time these days.)

**Comment by cy89** [comment]
@rmlarsen is this something you've got an opinion about? Seems like it might be about deterministic numerics.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @cy89: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @madrugado 
We see that you are using old version of Tensorflow which is officially considered as end of life, We recommend that you upgrade to 2.5  version and let us know if the issue still persists in newer versions .Please open a new issue in case you face any errors, we will get you the right help .Thanks!
